include/PolygonSimplification.hpp

```diff
@@ -33,62 +33,107 @@
-						}
-						*/
-						
-						CGAL::Container_from_circulator<PolyExplorer::Halfedge_around_face_const_circulator> edges(face.halfedge());
-						using Edge = BareType<decltype(*edges.begin())>;
-						vector<CGALPoint> points;
-						// As long as we are using a Bounded_kernel, we don't need to worry that the vertex is actually a ray.
-						std::transform(edges.begin(), edges.end(), std::back_inserter(points), [](const Edge &edge){ return edge.vertex()->point(); });
-						
-						return f(points);
-					});
-				}
-				
-				void debugNef(const NefPolyhedron &poly) {
-					vector<CGALPolygon> temp;
-					PolyExplorer explorer = poly.explorer();
-					extractFiniteFaces(poly.explorer(), temp, points2Poly);
-					std::for_each(temp.cbegin(), temp.cend(), [](const CGALPolygon &poly){
-						std::cout << "\t" << poly << " ( is simple? " << poly.is_simple() << " )" << std::endl;
-					});
-				}
-			}
+			static constexpr auto identity = [](std::vector<CGALPoint> &p) constexpr { return p; };
+			static constexpr auto points2Poly = [](std::vector<CGALPoint> &p) constexpr { return CGALPolygon(p.cbegin(), p.cend()); };
+			template<typename T> using BareType = typename std::remove_cv<typename std::remove_reference<T>::type>::type;
 			
-			std::vector<CGALPolygon> simplifyPolygon(const CGALPolygon &inp, bool auto_close = true){
+			/**************************************************
+			 * Extracts the finite faces from a Nef polyhedron
+			 * @arg explorer The `CGAL::Nef_polyhedron_2<CGALBoundedKernel>::Explorer
+			 * to traverse.
+			 * @arg faces The output container, must support
+			 * `std::back_inserter`. Typically `std::vector<E>
+			 * for some face-type `E`.
+			 * @arg f A functor from 
+			 * `std::vector<CGAL::Exact_predicates_exact_constructions_kernel::Point_2>` 
+			 * to `E`, where `E` is the element type of `faces`.
+			 **************************************************/
+			template<typename T, typename F = decltype(identity)> static void extractFiniteFaces(const PolyExplorer &explorer, T& faces, F f = identity) {
+				using Face = BareType<decltype(*explorer.faces_begin())>;
+				std::transform(++(explorer.faces_begin()), explorer.faces_end(), std::back_inserter(faces), [&f](const Face &face){
+					/*
+					if (std::distance(face.fc_begin(), face.fc_end()) > 0) {
+						std::cerr << ("There shouldn't be a hole in the interior of this face, "
+									  "because it's the result of tracing a single polyline into the plane\n") << std::endl;
+					}
+					*/
+					
+					CGAL::Container_from_circulator<typename PolyExplorer::Halfedge_around_face_const_circulator> edges(face.halfedge());
+					using Edge = BareType<decltype(*edges.begin())>;
+					std::vector<CGALPoint> points;
+					// As long as we are using a Bounded_kernel, we don't need to worry that the vertex is actually a ray.
+					std::transform(edges.begin(), edges.end(), std::back_inserter(points), [](const Edge &edge){ return edge.vertex()->point(); });
```

Comment:
FWIW, I'm pretty sure that the points associated with each vertex will be `CGALKernel::Point_2` and not `CGALPoint`, so this is not going to preserve IDs. If this doesn't work for you, you're going to need to provide a custom kernel that uses your custom points.

Reply:
@elfprince13 what makes you say that?

Reply:
Because `CGAL::Nef_polyhedron_2` is templated on the kernel type and not the point type, so the internal data structures have no way of knowing about the ids. In _principle_, that doesn't preclude an interface based on some kind of type-erasure mechanics, but I know enough about CGAL's zero-cost abstractions design philosophy to tell you that's not what's happening here.

Have you actually run tests to confirm if a no-op simplification on an already simple polygon preserves IDs?

Reply:
No, I have not, but that certainly sounds *de rigueur*.